api/websocket.go

```diff
@@ -70,7 +104,8 @@ func (wsSvr *WebsocketHandler) handleConnection(ctx context.Context, ws *websock
 	})
 
 	ctx, cancel := context.WithCancel(ctx)
-	go ping(ctx, ws, cancel)
+	safeWs := &safeWebsocketConn{ws: ws}
+	go ping(ctx, safeWs, cancel)
```

Comment:
1. according to https://pkg.go.dev/github.com/gorilla/websocket#hdr-Concurrency, Reader method also need concurrent control, like `SetReadDeadline`,  `SetReadLimit`
2. L95-103 above, `SetReadDeadline` called 2 times, is that a duplicate?

Reply:
1. see websocket code, panic is only triggered on write(WriteMessage, WriteJSON), Set* will not
```go
	if c.isWriting {
		panic("concurrent write to websocket connection")
	}
```

2. the first one is used to set the initial read deadline, ensuring that data is read from the WebSocket connection within the specified time. The second one is used to reset the read deadline after receiving a heartbeat message.

Reply:
but pkg.go.dev clearly stated:

"Applications are responsible for ensuring that no more than one goroutine calls the write methods (NextWriter, SetWriteDeadline, WriteMessage, WriteJSON, EnableWriteCompression, SetCompressionLevel) concurrently and **that no more than one goroutine calls the read methods** (NextReader, SetReadDeadline, ReadMessage, ReadJSON, SetPongHandler, SetPingHandler) concurrently."